client/ayon_shotgrid/plugins/publish/collect_shotgrid_entities.py

```diff
@@ -0,0 +1,223 @@
+
+            ayon_asset = instance.data.get("assetEntity")
+            sg_asset_id = ayon_asset.get("data", {}).get("shotgridId")
+
+            sg_asset = None
+
+            if sg_asset_id:
+                sg_asset = sg_assets_by_id.get(int(sg_asset_id))
+
+            instance.data["shotgridEntity"] = sg_asset
+            self.log.info(
+                "Collected Instance Shotgrid Asset: {}".format(
+                    instance.data["shotgridEntity"]
+                )
+            )
+
+            sg_task = None
+            ayon_task_name = instance.data.get("task")
+
+            if ayon_task_name:
+                ayon_task = self._get_ayon_task(
+                    context.data["projectName"],
+                    ayon_task_name,
+                    ayon_asset["_id"]
+                )
+
+                if ayon_task:
+                    sg_task_id = ayon_task.get("attrib", {}).get("shotgridId")
+
+                    if sg_task_id:
+                        sg_task = sg_tasks_by_id.get(
+                            int(sg_task_id)
+                        )
+
+            instance.data["shotgridTask"] = sg_task
+            self.log.info(
+                "Collected Instance Shotgrid Task: {}".format(
+                    instance.data["shotgridTask"]
+                )
+            )
+
+    def _get_ayon_task(self, project_name, task_name, folder_id):
+        """Helper method to get the Ayon Task by it's name and parent folder.
+
+        Args:
+            project_name (str): The Ayon project name.
+            task_name (str): The task name we are looking for.
+            folder_id (str): The ID of the parent folder.
+
+        Returns:
+            ayon_task (dict): The Task dict representation if found.
+        """
+        ayon_task = {}
+
+        for task in ayon_api.get_tasks(
+            project_name,
+            folder_ids=[folder_id]
+        ):
+            if task.get("name") == task_name:
+                ayon_task = task
+                break
+
+        return ayon_task
+
+    def _get_sg_tasks_by_id(self, context):
+        """ Get all instances tasks from Shotgrid.
+
+        We loop over all the instances and gather all the tasks so we can
+        perform one Shotgird query and get all Tasks, then build a dict by their
+        IDs.
+
+        Args:
+            context (pyblish.Plugin.context): The current publish context.
+
+        Returns:
+            sg_tasks_by_id (dict): All the found tasks, with their ID as key
+                and the Shotgrid entity dict as value.
+        """
+        sg_session = context.data["shotgridSession"]
+        sg_tasks_ids = []
+        sg_tasks_by_id = {}
+
+        for instance in context:
+            ayon_asset = instance.data.get("assetEntity")
+            ayon_task_name = instance.data.get("task")
+
+            if not ayon_task_name:
+                continue
+
+            ayon_task = self._get_ayon_task(
```

Comment:
You should rather collect tasks by all available folder ids once, and use that for rest of plugin. Right now you're querying ayon tasks one by one, twice.